File: java/doit_java/Chapter5/src/exercise/MyDate.java
/*
Q4
05-7에서 예제로 나온 MyDate와 MyDateTest 클래스를 완성해 봅시다.

 <MyDate 클래스 완성하기>
 - day, month, year 변수는 private으로 선언합니다.
 - 각 변수의 get, set 메서드를 public으로 만듭니다.
 - MyDate(int day, int month, int year) 생성자를 만듭니다.
 - public boolean isValid() 메서드를 만들어 날짜가 유효한지 확인합니다.
 - MyDateTest 클래스에서 생성한 MyDate 날짜가 유효한지 확인합니다.
 */
package exercise;

import java.util.Calendar;
import static java.util.Calendar.getInstance;

public class MyDate {
    private int day;
    private int month;
    private int year;
    private boolean isValid = true;

    public MyDate(int day, int month, int year) {
        setYear(year);
        setMonth(month);
        setDay(day);
    }
    public int getDay() {
        return day;
    }
    public void setDay(int day) {
        switch (month) {
            case 1: case 3: case 5: case 7: case 8: case 10: case 12:
                if (day < 0 || day > 31)
                    isValid = false;
                else
                    this.day = day;
                break;
            case 4: case 6: case 9: case 11:
                if (day < 0 || day > 30)
                    isValid = false;
                else
                    this.day = day;
                break;
            case 2:
                if ((year % 4 == 0 && year % 100 != 0) || year % 400 == 0) { // 윤년인 경우
                    if (day < 0 || day > 29)
                        isValid = false;
                    else
                        this.day = day;
                } else {
                    if (day < 0 || day > 28)
                        isValid = false;
                    else
                        this.day = day;
                }
                break;
            default:
                isValid = false;
        }
    }
    public int getMonth() {
        return month;
    }
    public void setMonth(int month) {
        if (month < 1 || month > 12)
            isValid = false;
        else
            this.month = month;
    }
    public int getYear() {
        return year;
    }
    public void setYear(int year) {
        if (year > getInstance().get(Calendar.YEAR))
            isValid = false;
        else
            this.year = year;
    }
    public String isValid() {
        if (isValid)
            return "유효한 날짜입니다.";
        else
            return "유효하지 않은 날짜입니다.";
    }
}

File: python/p_oop2/chap05/excercise/my_date.py
class MyDate:
    def __init__(self, day, month, year):
        self.is_valid = True
        self.set_year(day)
        self.set_month(month)
        self.set_day(day)

    def get_day(self):
        return self.day

    def set_day(self, day):
        month_30 = [4, 6, 9, 11]
        month_31 = [1, 3, 5, 7, 8, 10, 12]
        if self.month in month_31:
            if 0 < day <= 31:
                self.day = day
            else:
                self.is_valid = False
        elif self.month in month_30:
            if 0 < day < 31:
                self.day = day
            else:
                slef.is_valid = False
        elif self.month == 2:
            if (self.year % 4 == 0 and self.year % 100 != 0) or self.year % 400 == 0:
                if 0 < day <= 29:
                    this.day = day
                else:
                    self.is_valid = False
            else:
                if 0 < day < 29:
                    self.day = day
                else:
                    self.is_valid = False
        else:
            is_valid = False

    def get_month(self):
        return self.month

    def set_month(self, month):
        if 0 < month <= 12:
            self.month = month
        else:
            self.is_valid = False

    def get_year(self):
        return self.year

    def set_year(self, year):
        if year > 0:
            self.year = year
        else:
            self.is_valid = False

    def is_valid_date(self):
        if self.is_valid:
            return "유효한 날짜입니다."
        else:
            return "유효하지 않은 날짜입니다."


def main():
    date1 = MyDate(29, 2, 2000)
    print(date1.is_valid_date())

    date2 = MyDate(2, 10, 2006)
    print(date2.is_valid_date())


if __name__ == '__main__':
    main()
